Documentation/coding-guidelines/interop-pinvokes.md

```diff
@@ -0,0 +1,239 @@
+
+```C#
+[StructLayout(LayoutKind.Sequential, CharSet = CharSet.Unicode)]
+public struct UnicodeCharStruct
+{
+    public char c;
+}
+```
+
+`string` is blittable if it isn't contained in another type and it's being passed as an argument that is marked with `[MarshalAs(UnmanagedType.LPWStr)]` or the `[DllImport]` has `CharSet = CharSet.Unicode` set.
+
+You can see if a type is blittable by attempting to create a pinned `GCHandle`. If the type is not a string or considered blittable `GCHandle.Alloc` will throw an `ArgumentException`.
+
+
+[Blittable and Non-Blittable Types](https://msdn.microsoft.com/en-us/library/75dwhxf7.aspx "MSDN")  
+[Default Marshalling for Value Types](https://msdn.microsoft.com/en-us/library/0t2cwe11.aspx "MSDN")
+
+Keeping Managed Objects Alive
+-----------------------------
+`GC.KeepAlive()` will ensure an object stays in scope until the KeepAlive method is hit.
+
+[`HandleRef`][6] allows the marshaller to keep an object alive for the duration of a P/Invoke. It can be used instead of `IntPtr` in method signatures. `SafeHandle` effectively replaces this class and should be used instead.
+
+[6]: https://msdn.microsoft.com/en-us/library/system.runtime.interopservices.handleref.aspx "MSDN"
+
+[`GCHandle`][7] allows pinning a managed object and getting the native pointer to it. Basic pattern is:  
+
+``` C#
+GCHandle handle = GCHandle.Alloc(obj, GCHandleType.Pinned);
+IntPtr ptr = handle.AddrOfPinnedObject();
+handle.Free();
+```
+
+[7]: https://msdn.microsoft.com/en-us/library/system.runtime.interopservices.gchandle.aspx "MSDN"
+
+Pinning is not the default for `GCHandle`. The other major pattern is for passing a reference to a managed object through native code back to managed code (via a callback, typically). Here is the pattern:
+
+``` C#
+GCHandle handle = GCHandle.Alloc(obj);
+SomeNativeEnumerator(callbackDelegate, GCHandle.ToIntPtr(handle));
+
+// In the callback
+GCHandle handle = GCHandle.FromIntPtr(param);
+object managedObject = handle.Target;
+```
+
+Structs
+-------
+
+Managed structs are created on the stack and aren't removed until the method returns. By definition then, they are "pinned" (it won't get moved by the GC). You can also simply take the address in unsafe code blocks if native code won't use the pointer past the end of the current method. 
+
+Class fields in structs are marshalled as pointers (`IntPtr`) _unless_ the class has explicit layout, in which case the class fields are embedded in the marshalling buffer. For example:
+
+``` C#
+public struct MyStruct
+{
+    public class MyClass;
+}
+```
+
+`Marshal.SizeOf<MyStruct>()` will return `sizeof(IntPtr)` unless `MyClass` has explicit layout (via `[StructLayout]`). If `MyClass` has explicit layout `Marshal.SizeOf<MyStruct>()` will return `Marshal.SizeOf<MyClass>()` in this case (as `MyStruct` has no other fields).
```

Comment:
`Marshal.SizeOf` is pretty slow API. It should be avoided.

If you are using blittable structs, you can just use `sizeof(MyStruct)` that the JIT compiles into a constant.